sdk/iot/Azure.Iot.Hub.Service/tests/DevicesClientTests.cs

```diff
@@ -2,66 +2,393 @@
+
+        /// <summary>
+        /// Test bulk device and twin creation.
+        /// </summary>
+        [Test]
+        public async Task DevicesClient_BulkCreation_DeviceWithTwin()
+        {
+            string testDeviceprefix = $"bulkDeviceWithTwin";
+            string userPropertyName = "user";
+            string userPropertyValue = "userA";
+
+            IoTHubServiceClient client = GetClient();
+
+            IDictionary<string, object> desiredProperties = new Dictionary<string, object>
+            {
+                { userPropertyName, userPropertyValue }
+            };
+
+            // We will build multiple devices and all of them with the same desired properties for convenience.
+            IDictionary<DeviceIdentity, TwinData> devicesAndTwins = BuildDevicesAndTwins(testDeviceprefix, BULK_DEVICE_COUNT, desiredProperties);
+
+            try
+            {
+                // Create all devices with twins
+                Response<BulkRegistryOperationResponse> createResponse = await client.Devices.CreateIdentitiesWithTwinAsync(devicesAndTwins).ConfigureAwait(false);
+
+                Assert.IsTrue(createResponse.Value.IsSuccessful, "Bulk device creation ended with an error");
+
+                // Verify that the desired properties were set
+                // For quicker test run, we will only verify the first device on the list.
+                Response<TwinData> getResponse = await client.Devices.GetTwinAsync(devicesAndTwins.Keys.First().DeviceId).ConfigureAwait(false);
+                getResponse.Value.Properties.Desired[userPropertyName].Should().Be(userPropertyValue);
+            }
+            finally
+            {
+                await Cleanup(client, devicesAndTwins.Keys);
+            }
+        }
+
+        /// <summary>
+        /// Test query by getting all twins.
+        /// For the purpose of this test, we will create multiple devices (and device twins as a byproduct)
+        /// and list all twins and verify the query returns everything expected.
+        /// </summary>
+        [Test]
+        public async Task DevicesClient_Query_GetTwins()
+        {
+            string testDeviceprefix = $"bulkDevice";
+
+            IEnumerable<DeviceIdentity> devices = BuildMultipleDevices(testDeviceprefix, BULK_DEVICE_COUNT);
+
+            IoTHubServiceClient client = GetClient();
+
+            try
+            {
+                // Create all devices
+                Response<BulkRegistryOperationResponse> createResponse = await client.Devices.CreateIdentitiesAsync(devices).ConfigureAwait(false);
+
+                Assert.IsTrue(createResponse.Value.IsSuccessful, "Bulk device creation must be successful");
+
+                // We will retry the operation since it can take some time for the query to match what was recently created.
+                int matchesFound = 0;
+                DateTimeOffset startTime = DateTime.UtcNow;
+
+                while (DateTime.UtcNow - startTime < _queryMaxWaitTime)
```

Comment:
I'm a little confused. Is this loop here because the devices are not immediately queryable after the bulk creation?

Reply:
If so, that may be worth adding a comment for

Reply:
I am sorry, I am going to have to disagree, I think comments in code are useful for future authors of the code. They may seem redundant, but I don't see any harm in having step by step comments in the code even though the action can be identified by the code. if we follow that logic, all comments in code are redundant. am I making sense? One can argue that code comments are suppose to explain the reasoning behind the logic that was employed in the code, but they are also suppose to help follow what is happening in the block of code.

Reply:
You are correct. There are comments on line 304